shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -227,20 +227,34 @@
   }
   UIView* flutter_view = flutter_view_.get();
 
-  // This can be more efficient, instead of removing all views and then re-attaching them,
-  // we should only remove the views that has been completly removed from the layer tree, and
-  // reorder the views using UIView's bringSubviewToFront.
-  // TODO(amirh): make this more efficient.
-  // https://github.com/flutter/flutter/issues/23793
-  for (UIView* sub_view in [flutter_view subviews]) {
-    [sub_view removeFromSuperview];
+  std::unordered_set<int64_t> composition_order_set;
+  for (int64_t view_id : composition_order_) {
+    composition_order_set.insert(view_id);
+  }
+
+  for (int64_t view_id : active_composition_order_) {
+    if (composition_order_set.find(view_id) == composition_order_set.end()) {
+      [touch_interceptors_[view_id].get() removeFromSuperview];
+      [overlays_[view_id]->overlay_view.get() removeFromSuperview];
+    }
   }
 
+  composition_order_set.clear();
   active_composition_order_.clear();
+
   for (size_t i = 0; i < composition_order_.size(); i++) {
     int view_id = composition_order_[i];
-    [flutter_view addSubview:touch_interceptors_[view_id].get()];
-    [flutter_view addSubview:overlays_[view_id]->overlay_view.get()];
+    if (touch_interceptors_[view_id].get().superview == flutter_view) {
```

Comment:
Let's use the same condition for both the overlay and the interceptor.
They should always be attached/detached together, we can even assert that the superview for both is the same just before the condition.

Reply:
done

Reply:
done